kratos/future/solving_strategies/builders/builder.h

```diff
@@ -0,0 +1,372 @@
+                    eff_dof_addition_fn(i_dof);
+                }
+            }
+
+            // Loop the constraints to add the slave DOFs
+            // Note that we assume that there constraints are always one to many (not many to many)
+            auto slave_dof_addition_fn = [&](IndexType Index){
+                auto it_constraint = mpModelPart->MasterSlaveConstraintsBegin() + Index;
+                const auto& r_masters = it_constraint->GetMasterDofsVector(); //TODO: we don't use TLS for this (decide what to do once we finish the parallelism discussion)
+                const IndexType slave_eq_id = (it_constraint->GetSlaveDofsVector()[0])->EquationId();
+                for (const auto& rp_master_dof : r_masters) {
+                    const IndexType master_eff_eq_id = rp_master_dof->EffectiveEquationId();
+                    rConstraintsSparseGraph.AddEntry(slave_eq_id, master_eff_eq_id);
+                }
+            };
+
+            if constexpr (TSparseGraphType::IsThreadSafe) {
+                IndexPartition<IndexType>(n_constraints).for_each(slave_dof_addition_fn);
+            } else {
+                for (IndexType i_constraint = 0; i_constraint < n_constraints; ++i_constraint) {
+                    slave_dof_addition_fn(i_constraint);
+                }
+            }
+        }
+    }
+
+    /**
+     * @brief Applies the constraints to the linear system of equations
+     * This method applies the linear system constraints to the linear system of equations
+     * Note that linear system constraints includes both the master-slave and the Dirichlet
+     * constraints. The way the constraints are applied will be reimplemented and applied
+     * in the derived classes depending on the build type.
+     * @param rLinearSystemContainer Auxiliary container with the linear system arrays
+     */
+    virtual void ApplyLinearSystemConstraints(LinearSystemContainer<TSparseMatrixType, TSystemVectorType>& rLinearSystemContainer)
+    {
+        KRATOS_ERROR << "Calling base class 'ApplyLinearSystemConstraints'." << std::endl;
+    }
+
+    /**
+     * @brief Set the solution values vector
+     * This method calculates the complete solution vector (x) from the effective DOF set
+     * For doing so the effective DOF values are retrieved to then apply the constraints
+     * by doing x = T * y + q, being T and q the constraints relation matrix and constant
+     * vector respectively and y the auxliary vector containing the effective DOFs values
+     * Note that the vector y already includes the Dirichlet constraints as values are
+     * retrieved from the DOF (i.e., nodal) database.
+     * @param rEffectiveDofSet The effective DOFs array (i.e., those that are not slaves)
+     * @param rConstraintsRelationMatrix The constraints relation matrix
+     * @param rConstraintsConstantVector The constraints constant vector
+     * @param rSolutionVector The solution vector
+     */
+    void CalculateSolutionVector(
```

Comment:
This shouldn't be in the scheme? I mean, scheme does the large inter matrix operations, no?

Reply:
This mainly does the u = T*y+b operation and it is the builder the one applying the constraints. This was my rationale for doing it like this.